src/AzureExtension/DevBox/Helpers/WingetConfigWrapper.cs

```diff
@@ -182,56 +207,61 @@ private void SetStateForCustomizationTask(TaskJSONToCSClasses.BaseClass response
 
                     // If the status is waiting for a user session, there is no need to check for other
-                    // individual task statuses. We add a wait since Winget takes time to start applying
-                    // the configuration and we don't want to show the same message immediately after.
-                    if (responseStatus == "WaitingForUserSession")
-                    {
-                        ApplyConfigurationActionRequiredEventArgs eventArgs = new(new WaitingForUserAdaptiveCardSession(_resumeEvent));
-                        ActionRequired?.Invoke(this, eventArgs);
-                        WaitHandle.WaitAny(new[] { _resumeEvent });
-                        Thread.Sleep(TimeSpan.FromSeconds(30));
-                        break;
-                    }
-                    else if (_oldUnitState[i] != unitState)
-                    {
+                    // individual task statuses.
+                    if (responseStatus == "WaitingForUserSession")
+                    {
+                        isWaitingForUserSession = true;
+                        continue;
+                    }
+
+                    if (responseStatus == "Running")
+                    {
+                        isAnyTaskRunning = true;
+                    }
+
+                    if (_oldUnitState[i] != responseStatus)
+                    {
                         var task = _units[i];
-                        ConfigurationSetStateChangedEventArgs args = new(new(ConfigurationSetChangeEventType.UnitStateChanged, setState, unitState, null, task));
+                        var unitState = DevBoxOperationHelper.JSONStatusToUnitStatus(responseStatus);
+                        var resultInfo = responseStatus == "Failed" ? _commonfailureResult : null;
+                        ConfigurationSetStateChangedEventArgs args = new(new(ConfigurationSetChangeEventType.UnitStateChanged, setState, unitState, resultInfo, task));
                         ConfigurationSetStateChanged?.Invoke(this, args);
-                        _oldUnitState[i] = unitState;
+                        _oldUnitState[i] = responseStatus;
                     }
-
-                    _log.Information($"Unit Status: {unitState}");
+                }
+
+                // If waiting for user session and no task is running, show the adaptive card
+                // We add a wait since Dev Boxes take a little over 2 minutes to start applying
+                // the configuration and we don't want to show the same message immediately after.
+                if (isWaitingForUserSession && !isAnyTaskRunning)
+                {
+                    ApplyConfigurationActionRequiredEventArgs eventArgs = new(new WaitingForUserAdaptiveCardSession(_resumeEvent));
+                    ActionRequired?.Invoke(this, eventArgs);
+                    WaitHandle.WaitAny(new[] { _resumeEvent });
+                    Thread.Sleep(TimeSpan.FromSeconds(135));

```

Comment:
I'm thinking we could do` Thread.Sleep(TimeSpan.FromMinutes(2.5))` what do you think?

Reply:
TIL. But I want it to be 2 minutes and 15 seconds.

Reply:
I hope we get a way to check on status from DevBox later. Having such a big wait doesn't look good.
Also, what would happen if user closed DevHome? Do we need a cancellation token to break this wait?

Reply:
This wait is very much intentional on Dev Box's side. The team needs to make sure that the required services are running and the required folders have been created. They are looking into making this wait shorter for non-first-time-logins.

And since there isn't a way to stop set tasks, I haven't added any cancellation.